lib/services/local/users.go

```diff
@@ -148,49 +168,60 @@ func (s *IdentityService) CreateUser(user types.User) error {
 		Expires: user.Expiry(),
 	}
 
-	if _, err = s.Create(context.TODO(), item); err != nil {
-		return trace.Wrap(err)
+	lease, err := s.Create(ctx, item)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
 
 	if auth := user.GetLocalAuth(); auth != nil {
-		if err = s.upsertLocalAuthSecrets(user.GetName(), *auth); err != nil {
-			return trace.Wrap(err)
+		if err = s.upsertLocalAuthSecrets(ctx, user.GetName(), *auth); err != nil {
+			return nil, trace.Wrap(err)
 		}
 	}
-	return nil
+	user.SetRevision(lease.Revision)
+	return user, nil
 }
 
 // UpdateUser updates an existing user.
 func (s *IdentityService) UpdateUser(ctx context.Context, user types.User) error {
+	_, err := s.UpdateUserWithContext(ctx, user)
+	return trace.Wrap(err)
+}
+
+// UpdateUserWithContext updates an existing user.
+// TODO(tross) remove this once oss and e are converted to using the new signature.
+func (s *IdentityService) UpdateUserWithContext(ctx context.Context, user types.User) (types.User, error) {
 	if err := services.ValidateUser(user); err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
 	// Confirm user exists before updating.
 	if _, err := s.GetUser(user.GetName(), false); err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
 	value, err := services.MarshalUser(user.WithoutSecrets().(types.User))
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 	item := backend.Item{
-		Key:     backend.Key(webPrefix, usersPrefix, user.GetName(), paramsPrefix),
-		Value:   value,
-		Expires: user.Expiry(),
-		ID:      user.GetResourceID(),
+		Key:      backend.Key(webPrefix, usersPrefix, user.GetName(), paramsPrefix),
+		Value:    value,
+		Expires:  user.Expiry(),
+		ID:       user.GetResourceID(),
+		Revision: user.GetRevision(),
 	}
-	_, err = s.Update(ctx, item)
+	lease, err := s.Update(ctx, item)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 	if auth := user.GetLocalAuth(); auth != nil {
-		if err = s.upsertLocalAuthSecrets(user.GetName(), *auth); err != nil {
-			return trace.Wrap(err)
+		if err = s.upsertLocalAuthSecrets(ctx, user.GetName(), *auth); err != nil {
+			return nil, trace.Wrap(err)
 		}
 	}
-	return nil
+	user.SetRevision(lease.Revision)
```

Comment:
Should we zero the ID?

Reply:
Probably not, but of all the possible values, 0 is less offensive than "whatever was in the resource that the caller passed in".

Reply:
In `Get` we set the resource ID at unmarshal time, from the actual ID returned by the backend. We don't have the ID from the lease here, so we either leave whatever random ID was in the resource passed by the caller or we set it to 0, and I think that the latter is preferrable. We could also set some other sentinel value like `0x7fff_ffff_ffff_ffff` if you prefer.